### ISSUE TITLE
Add useRouter Hook

### ISSUE THREAD

**Issue by evdama** [ReactTraining/react-router#6430]
Hi guys,

at current React Conf [Hooks](https://reactjs.org/docs/hooks-intro.html) have been announced for React v16.7. What's the plan for supporting them? Will there be a `useRouter()` any time soon?

https://reactjs.org/docs/hooks-faq.html#what-do-hooks-mean-for-popular-apis-like-redux-connect-and-react-router

**Comment by pshrmn** [comment]
While I don't know any details about what Michael is planning, _when_ React Router supports them largely depends on the timeline of how fast React 16.7 is released.

React Router currently has its `4.4` beta out, which will probably become final sooner than 16.7 becomes final. In that case, hooks would be a `4.5` feature. RR can put out an alpha/beta build of `4.5`, but it shouldn't become "official" until hooks get a final release.

**Comment by third774** [comment]
It looks like `__RouterContext` is exposed in `v4.4.0-beta.5` - so you could probably just do something like:

```js
import {useContext} from "react"
export default function useRouter() {
  return useContext(__RouterContext)
}
```

Might be a horrible idea though. 🤷‍♂️

**Comment by timdorr** [comment]
Yes, we don't intend to make our context instance public. You should use our public APIs to access it, which allows us to warn you when it's used incorrectly and things like that.

**Comment by third774** [comment]
You could also create your own wrapper to get access:

```js
import React from "react";
import { BrowserRouter, Route } from "react-router-dom";

export const RouterContext = React.createContext({});

const CustomBrowserRouter = ({ children }) => (
  <BrowserRouter>
    <Route>
      {(routeProps) => (
        <RouterContext.Provider value={routeProps}>
          {children}
        </RouterContext.Provider>
      )}
    </Route>
  </BrowserRouter>
);

export default CustomBrowserRouter;
```

Then wrap this around your App instead of the normal `<BrowserRouter />`

```js
import React from "react";
import CustomBrowserRouter from "./CustomBrowserRouter";

function App() {
  return (
    <CustomBrowserRouter>
      <div className="App">
        {/* The rest of your app  */}
      </div>
    </CustomBrowserRouter>
  );
}

export default App;
```

Then you can make a useRouter hook:
```js
import {useContext} from "react"
import {RouterContext} from "./CustomBrowserRouter"

export default function useRouter() {
  return useContext(RouterContext)
}
```

**Comment by FredyC** [comment]
@timdorr But in that case, you need to export custom hook that would wrap around the context because `useContext(Consumer)` is not supported way.

**Comment by pshrmn** [comment]
@FredyC RR can export a hook (`useRouter`) to provide its context data without the context being public. I only mentioned the consumer to note that the currently exported context can theoretically be removed, which would break the code of anyone relying on it.

**Comment by mjackson** [comment]
Ya, I think a `useRouter` hook makes a lot of sense. It'll probably work just like a path-less `<Route>` currently does, and just pass along the current `match` object at that point in the tree.

We won't be able to include this feature in the 4.x branch though without shipping some kind of hooks polyfill, which I'd rather not do. We should be able to safely assume that anybody using hooks is also using React 16.7+ which will be our dependency in version 5.

**Comment by Janpot** [comment]
@mjackson can't you do something like

```jsx
import React, { useContext } from 'react';

export function useRouter () {
  if (!useContext) {
    throw new Error('React >= 16.7 required');
  }
  return useContext(...);
}
```

No polyfill required, you just can't use `useRouter` unless you're on `16.7` or higher. This means it can go on a `react-router` minor version.

**Comment by xiaoxiangmoe** [comment]
```ts

import { useContext }from 'react'
import { RouteComponentProps, StaticContext } from 'react-router';
import * as H from 'history';

const RouterContext =  require('react-router').__RouterContext;

// FIXME:  use official API when https://github.com/ReactTraining/react-router/pull/6453 merged

export function useRouter<Params extends { [K in keyof Params]?: string } = {}, C extends StaticContext = StaticContext, S = H.LocationState>() {
  return useContext(RouterContext) as RouteComponentProps<Params,C,S>
}
```

**Comment by FredyC** [comment]
@mjackson I am doing the same since you don't want to release version with `unstable_useRouter` for whatever reasons :) It's too good to live without that anymore :)

**Comment by panjiesw** [comment]
I also use the same workaround with @xiaoxiangmoe. Of course it's only temporary, as noted in the snippet, will replace it with `unstable_useRouter`, and eventually, hopefully `useRouter` as the official way.

**Comment by agilgur5** [comment]
Used the same workaround initially since `unstable_useRouter` still has yet to be released, but then I found https://github.com/CharlesStover/use-react-router which does the workaround for you, allowing `const {history, match, location} = useReactRouter()`. Figured I should post it here in case it might help anyone because I only found it a few days after implementing the workaround manually (surprised no one else has linked it!)

It does implement pub-sub, which may be different from expected behavior; the implementation details are described in https://medium.com/@Charles_Stover/how-to-convert-withrouter-to-a-react-hook-19bb02a29ed6

**Comment by mateusznowosielski** [comment]
Not sure if anyone run into the same issue but when I run it in a browser, `useContext(__RouterContext)` returns `{ match, location, history }` objects correctly, but when I'm running my tests and mocking my router, let's say `<MemoryRouter initialEntries={['/something/here']}>` then `location` and `history` object are correct but `match` is always `{ path: '/', url: '/', params: {}, isExact: false }`. It doesn't represent what it should be.

**Comment by FredyC** [comment]
@mateusznowosielski Shouldn't you also specify `initialIndex`? Either way, you can try `<Route>{({ match }) => ...</Route>` just to see if it's working. I don't think it's a hook related.

**Comment by mateusznowosielski** [comment]
@FredyC `initialIndex` wasn't necessary when I had only 1 `initialEntry`, but I had to wrap my component within `<Route component={ui} path={path} />`. Thanks for the suggestion.

**Comment by ricokahler** [comment]
Instead of a `useRouter` hook, could we consider a `useRoute` instead? I've always been bothered that `withRouter` doesn't allow passing in a path.

Would anyone be opposed to `useRoute` like so:

```js
import { useRoute } from 'react-router';

export default function SomeComponent() {
  const { match } = useRoute({ path: '/:something/:to/:match' });
}
```

The idea is that the `useRoute` would take in all the props a `<Route />` would take in and pass them to the second argument of `matchPath`.

@mjackson, what do you think?

**Comment by FredyC** [comment]
@ricokahler There is already a [PR that does that](https://github.com/ReactTraining/react-router/pull/6453).  It just needs a little update as @mjackson managed to stall it until hooks became an official :D

**Comment by ricokahler** [comment]
@FredyC I did not realize the PR already did this! I think I just assumed since `withRouter` didn't take any props that `useRouter` wouldn't either.

@mjackson  Forget I said anything and release this ha! 🎉

**Comment by flq** [comment]
Hello dear people - how is the "useRouter" situation?

**Comment by sibelius** [comment]
exporting react-router context api would make it easy to user build their own hooks on top of it

so, we don't need to wait until the final hook solution lands

**Comment by FredyC** [comment]
@sibelius There is a context exported in 4.4 beta ... https://github.com/ReactTraining/react-router/issues/6430#issuecomment-439056692

**Comment by AlexFrazer** [comment]
Here's what I had with typescript if it helps anyone viewing this thread:

First added some module augmentation to expose the `__RouterContext` as a type.

```ts
// types/react-router-dom.d.ts
import { RouteComponentProps } from 'react-router-dom';
import { Context } from 'react';

declare module 'react-router-dom' {
  export const __RouterContext: Context<RouteComponentProps<{}>>;
}
```

Then I made a custom hook around that.

```ts
import { useContext } from 'react';
import { __RouterContext, RouteComponentProps } from 'react-router-dom';

export default function useRouter<TParams = {}>() {
  return useContext(__RouterContext) as RouteComponentProps<TParams>;
}
```

And in use it would be like:
```ts
const { match: { params: { id } } } = useRouter<{ id: string }>();
```

**Comment by ms-ww** [comment]
Hi, do we still not have any implementation of hooks merged yet as of 5.0.0?

**Comment by sorahn** [comment]
@ms-ww 5.0 wasn't actually a feature release, just a synchronization of versions.

**Comment by dantman** [comment]
Looking at withRouter and stripping out the HOC stuff, it seems like a useRouter implementation would just be:

```js
import { useContext } from 'react';
import RouterContext from './RouterContext';
import invariant from 'tiny-invariant';

/**
 * A public hook to access the imperative API
 */
export default function useRouter() {
  invariant(useContext, 'React >= 16.8 required');

  const context = useContext(RouterContext);
  invariant(context, `You should not use useRouter outside a <Router>`);

  return context;
}
```

**Comment by sorahn** [comment]
@dantman That's basically what [this library](https://github.com/CharlesStover/use-react-router/blob/master/src/use-react-router.ts) is doing, but it has an added `forceUpdate` effect in it to make sure your buried component re-renders.

**Comment by tollycoast** [comment]
> You could also create your own wrapper to get access:
> 
> ```js
> import React from "react";
> import { BrowserRouter, Route } from "react-router-dom";
> 
> export const RouterContext = React.createContext({});
> 
> const CustomBrowserRouter = ({ children }) => (
>   <BrowserRouter>
>     <Route>
>       {(routeProps) => (
>         <RouterContext.Provider value={routeProps}>
>           {children}
>         </RouterContext.Provider>
>       )}
>     </Route>
>   </BrowserRouter>
> );
> 
> export default CustomBrowserRouter;
> ```
> 
> Then wrap this around your App instead of the normal `<BrowserRouter />`
> 
> ```js
> import React from "react";
> import CustomBrowserRouter from "./CustomBrowserRouter";
> 
> function App() {
>   return (
>     <CustomBrowserRouter>
>       <div className="App">
>         {/* The rest of your app  */}
>       </div>
>     </CustomBrowserRouter>
>   );
> }
> 
> export default App;
> ```
> 
> Then you can make a useRouter hook:
> 
> ```js
> import {useContext} from "react"
> import {RouterContext} from "./CustomBrowserRouter"
> 
> export default function useRouter() {
>   return useContext(RouterContext)
> }
> ```

This is a poor conclusion, it doesn't help anyone willing to learn from your explanation. "Then you can make a useRouter hook:" what is the purpose? Where is it used? What does it do?

**Comment by sarink** [comment]
I implemented a file called `routerHooks.ts` that looks like:
```
import { __RouterContext, RouteComponentProps } from 'react-router';
import { useContext } from 'react';

export const useLocation = () => {
  const context = useContext(__RouterContext);
  return context.location;
};

export function useMatch<P = {}>() {
  const context = useContext(__RouterContext) as RouteComponentProps<P>;
  return context.match;
}
```

**Comment by otakustay** [comment]
Our team has a set of more powerful router related hooks works fine in tens of projects:

```typescript
import {useContext, useMemo, useCallback} from 'react';
import {__RouterContext as RouterContext, RouteComponentProps} from 'react-router';
import {Location} from 'history';
import * as queryString from 'query-string';
import uriTemplate, {URITemplate} from 'uri-templates';

type ParsedQuery = queryString.ParsedQuery;

export const useRouter = <T>(): RouteComponentProps<T> => useContext(RouterContext) as RouteComponentProps<T>;

export const useLocation = (): Location => {
    const {location} = useRouter();
    return location;
};

export const useParams = <T>(): T => {
    const {match} = useRouter<T>();
    return match.params;
};

export const useQuery = <T extends ParsedQuery>(): T => {
    const {search} = useLocation();
    const query = useMemo(
        () => queryString.parse(search),
        [search]
    );
    return query as T;
};

interface UpdateQueryOptions {
    replace: boolean;
}

type UpdateQuery<T> = (patch: Partial<T>) => void;

type Visit<T> = (params: T) => void;

const USE_PUSH = {replace: false};

export const useUpdateQuery = <T extends ParsedQuery>(options: UpdateQueryOptions = USE_PUSH): UpdateQuery<T> => {
    const {history} = useRouter();
    const query = useQuery<T>();
    const {replace} = options;
    const updateQuery = useCallback(
        (patch: Partial<T>): void => {
            const newQuery = {...query, ...patch};
            const newSearch = queryString.stringify(newQuery);
            if (replace) {
                history.replace({search: newSearch});
            }
            else {
                history.push({search: newSearch});
            }
        },
        [history, query, replace]
    );
    return updateQuery;
};

export const useNavigate = <T>(to: string | URITemplate, options: UpdateQueryOptions = USE_PUSH): Visit<T> => {
    const {history} = useRouter();
    const {replace} = options;
    const template = useMemo(
        () => {
            if (typeof to === 'string') {
                return uriTemplate(to);
            }
            return to;
        },
        [to]
    );
    const visit = useCallback(
        (params: T): void => {
            const newLocation = template.fill(params as any);
            if (replace) {
                history.replace(newLocation);
            }
            else {
                history.push(newLocation);
            }
        },
        [template, history, replace]
    );
    return visit;
};
```

**Comment by daruj** [comment]
@otakustay what version of react-router are u using? I just installed 5.0 and I can't find `__RouterContext`

**Comment by Juansereina** [comment]
I guess this idea works if you are using <= 4.3.1,  but, it would be better to wait for this https://github.com/ReactTraining/react-router/pull/6453#issuecomment-474600561 because if something changes (https://github.com/ReactTraining/react-router/pull/6753)  then your code might breaks  


> Our team has a set of more powerful router related hooks works fine in tens of projects:
> 
> ```ts
> import {useContext, useMemo, useCallback} from 'react';
> import {__RouterContext as RouterContext, RouteComponentProps} from 'react-router';
> import {Location} from 'history';
> import * as queryString from 'query-string';
> import uriTemplate, {URITemplate} from 'uri-templates';
> 
> type ParsedQuery = queryString.ParsedQuery;
> 
> export const useRouter = <T>(): RouteComponentProps<T> => useContext(RouterContext) as RouteComponentProps<T>;
> 
> export const useLocation = (): Location => {
>     const {location} = useRouter();
>     return location;
> };
> 
> export const useParams = <T>(): T => {
>     const {match} = useRouter<T>();
>     return match.params;
> };
> 
> export const useQuery = <T extends ParsedQuery>(): T => {
>     const {search} = useLocation();
>     const query = useMemo(
>         () => queryString.parse(search),
>         [search]
>     );
>     return query as T;
> };
> 
> interface UpdateQueryOptions {
>     replace: boolean;
> }
> 
> type UpdateQuery<T> = (patch: Partial<T>) => void;
> 
> type Visit<T> = (params: T) => void;
> 
> const USE_PUSH = {replace: false};
> 
> export const useUpdateQuery = <T extends ParsedQuery>(options: UpdateQueryOptions = USE_PUSH): UpdateQuery<T> => {
>     const {history} = useRouter();
>     const query = useQuery<T>();
>     const {replace} = options;
>     const updateQuery = useCallback(
>         (patch: Partial<T>): void => {
>             const newQuery = {...query, ...patch};
>             const newSearch = queryString.stringify(newQuery);
>             if (replace) {
>                 history.replace({search: newSearch});
>             }
>             else {
>                 history.push({search: newSearch});
>             }
>         },
>         [history, query, replace]
>     );
>     return updateQuery;
> };
> 
> export const useNavigate = <T>(to: string | URITemplate, options: UpdateQueryOptions = USE_PUSH): Visit<T> => {
>     const {history} = useRouter();
>     const {replace} = options;
>     const template = useMemo(
>         () => {
>             if (typeof to === 'string') {
>                 return uriTemplate(to);
>             }
>             return to;
>         },
>         [to]
>     );
>     const visit = useCallback(
>         (params: T): void => {
>             const newLocation = template.fill(params as any);
>             if (replace) {
>                 history.replace(newLocation);
>             }
>             else {
>                 history.push(newLocation);
>             }
>         },
>         [template, history, replace]
>     );
>     return visit;
> };
> ```

**Comment by sdegutis** [comment]
I've been using this code successfully for a few months now:

```jsx
import * as React from "react";
import { BrowserRouter, Route } from "react-router-dom";

const RouterContext = React.createContext(null);

export const HookedBrowserRouter = ({ children }) => (
  <BrowserRouter>
    <Route>
      {(routeProps) => (
        <RouterContext.Provider value={routeProps}>
          {children}
        </RouterContext.Provider>
      )}
    </Route>
  </BrowserRouter>
);

export function useRouter() {
  return React.useContext(RouterContext);
}
```

But either way, I don't understand why this issue is still open? Is there a complication to adding any of these variations of useRouter code?

**Comment by maxparelius** [comment]
I also don't understand why there is still no `useRouter` hook that comes with `react-router` OOTB

**Comment by avkonst** [comment]
I moved from react-router to https://www.npmjs.com/package/hookrouter and very happy with it. Code becomes a lot simpler, easier to understand, shorter, size bundle is also smaller.

**Comment by japrogramer** [comment]
@avkonst how does that handle nested routes?

**Comment by avkonst** [comment]
Seems all right. There are docs with examples on nested routes. The only
component I added on top of hookrouter is Redirect (3 lines of code). This
allowed to replace react router completely.

**Comment by StringEpsilon** [comment]
@maxparelius : Shipping hooks would have meant dropping support for older react versions. That's quite a big breaking change. We'll probably get there in 6.0, but that's apparently quite the big release with more than just hooks. 

(And remember that 5.0 wasn't intended as a major release, so no intentional breaking changes like this made it in)

**Comment by MeiKatz** [comment]
@StringEpsilon Definitely, but we could introduce a Router hook with a version switch. What I mean, is, that for React versions lower than the needed version there will be an error, that this React version is not supported. For other React versions it should just work. This would be a compromise.

**Comment by StringEpsilon** [comment]
That seems reasonable.  Why aren't we doing that?

**Comment by MeiKatz** [comment]
Good question. I will push a proposal later that day.

**Comment by MeiKatz** [comment]
@StringEpsilon [Here](https://github.com/ReactTraining/react-router/pull/6812) we go!

**Comment by jayarjo** [comment]
I'm using react-router 5.0.1 here, does it mean I have the useRouter hook? Why this issue is even open if it's true?

**Comment by MeiKatz** [comment]
@jayarjo Because there is no `useRouter()` hook right now. My draft was rejected, because currently there seems to be no clear direction about how this hook should work and what it should return. There is also no clear opinion if there will be this hook at all or only hooks returning values from the context (e.g. params, location, history).

**Comment by mjackson** [comment]
We're not making any progress here, so I'm going to close this. We will have a few hooks in the next major release, maybe even in a minor release. If you absolutely cannot wait, please be my guest and use some other router that ships with hooks.

**Comment by dpikt** [comment]
@mjackson definitely understand that settling on hook API/behavior will take time and patience. However, can we keep this or a related issue open so folks can track when hooks have been added to the library? I often use Github's "closed" PR notifications to know when features are available, especially features that don't have a specific release designation.

**Comment by third774** [comment]
@dpikt — perhaps instead you could watch the repo for releases?

<img width="1037" alt="image" src="https://user-images.githubusercontent.com/8732191/60906656-41abe580-a23d-11e9-8287-eda1277aafdb.png">

**Comment by sdegutis** [comment]
While waiting what seems like an unusually long time for this Hooks-based API, it occurred to me that a few decisions React Router made have evolved in in a way that doesn't fully make sense for my use-cases. For example, you always get the full path when you create a route, instead of just the portion that the parent route didn't consume. And remembering that HTML5's History API isn't that hard, I figured I'd take a shot at a simplistic proof of concept at a simpler alternative. This is what I ended up with:

```tsx
import * as React from "react";

// The context is just the current path and navigation functions.
const RouterContext = React.createContext({
  path: '',
  pushPath: (path: string) => { },
  replacePath: (path: string) => { },
});

// This has to go around everything like in React Router.
function Router({ children }: { children: React.ReactNode }) {
  const [path, setPath] = React.useState(window.location.pathname);

  // Navigation functions.
  const pushPath = (newPath: string) => {
    if (newPath === path) return;
    window.history.pushState(null, '', newPath);
    setPath(newPath);
  }
  const replacePath = (newPath: string) => {
    if (newPath === path) return;
    window.history.replaceState(null, '', newPath);
    setPath(newPath);
  }

  // Navigation event handler that just updates React's state.
  function eventHandler() {
    setTimeout(() => {
      setPath(window.location.pathname);
    }, 0); // (MDN says this is probably needed)
  }

  // Listen to events as long as <Router> is mounted.
  React.useEffect(() => {
    window.addEventListener('popstate', eventHandler);
    return () => {
      window.removeEventListener('popstate', eventHandler);
    };
  }, []);

  // Pass down the initial path and navigation functions to everyone.
  return (
    <RouterContext.Provider value={{ path, pushPath, replacePath }}>
      {children}
    </RouterContext.Provider>
  );
}

// A simplified version of the React Router's <Route> component.
function Route({ path, render }: { path: string, render: () => JSX.Element }) {
  const context = React.useContext(RouterContext);

  // Render IFF this path matches the current path.
  if (!context.path.startsWith(path)) return null;

  // NB:
  // Remove the matched part of the path!
  // Only pass down the unmatched part.
  // Also pass down the navigation functions as-is.
  const newPath = context.path.substr(path.length);
  return (
    <RouterContext.Provider value={{ ...context, path: newPath }}>
      {render()}
    </RouterContext.Provider>
  );
}

// Simplistic demo of using the context to access the navigation functions.
// All you need is the built-in useContext hook! Nothing else needed.
function ButtonArea() {
  const context = React.useContext(RouterContext);
  return (<>
    <div>
      {/* This shows that you can access the unmatched portion fo the path. */}
      {/* If you want the full path, I suppose window.location.pathname works. */}
      {context.path}
    </div>
    <div>
      <button onClick={() => context.pushPath('/foo')}>/foo</button>
      <button onClick={() => context.pushPath('/foo/bar')}>/foo/bar</button>
      <button onClick={() => context.pushPath('/foo/qux')}>/foo/qux</button>
      <button onClick={() => context.pushPath('/foo/qux/bla')}>/foo/qux/bla</button>
    </div>
  </>)
}

// Further demo of using useContext to get the unconsumed path.
function DeepInnerRoute() {
  const context = React.useContext(RouterContext);
  return <b>bla ({context.path})</b>;
}

// The full demo.
// Notice the nested paths! This is *much* more intuitive to me.
export function NewRoutesTester() {
  return (
    <Router>
      <div>
        <Route path="/foo" render={() => <>
          <b>foo </b>
          <Route path="/bar" render={() => <>
            <b>bar </b>
          </>} />
          <Route path="/qux" render={() => <>
            <b>qux </b>
            <Route path="/bla" render={() => <DeepInnerRoute />} />
          </>} />
        </>} />
        <ButtonArea />
      </div>
    </Router>
  );
}
```

Obviously there's more features left to add, such as path-parameters, but these things are orthogonal to the basic architectural concepts laid out in this proof of concept.

That said, I'm not sure this architecture is compatible with React Router's current API or features, and it may not be something that can be integrated without a major version change, assuming it even matches the philosophy.

**Comment by kevlened** [comment]
@MeiKatz [submitted a PR for hooks that I found helpful](https://github.com/ReactTraining/react-router/pull/6812), so I copied the gist below. It wasn't merged [due to some uncertainty](https://github.com/ReactTraining/react-router/pull/6812#issuecomment-506940522) [about the api](https://github.com/ReactTraining/react-router/pull/6812#issuecomment-507926373), but these will work for me until `react-router` gets official hooks.

**Caveat**: it uses `__RouterContext`, which is a private api, so these hooks are unsupported.

```javascript
// react-router-hooks.js

import { useContext } from 'react';
import { __RouterContext as RouterContext } from 'react-router';

export function useRouter() {
    return useContext(RouterContext);
};

export function useParams() {
    const { match } = useRouter();
    return match.params;
}

export function useLocation() {
    const { location, history } = useRouter();
    
    function navigate(to, { replace = false } = {}) {
        if (replace) {
            history.replace(to);
        } else {
            history.push(to);
        }
    }
    
    return {
        location,
        navigate
    };
}
```

**Comment by jayarjo** [comment]
@sdegutis do you have it as a repo somewhere?

**Comment by sdegutis** [comment]
@jayarjo I don't currently have the luxury of maintaining an open source project, but here's a quick copy of the code I ended up with so far: https://github.com/sdegutis/simple-react-router. I'd love to see an API somewhat more like this for future React Router versions if possible.

**Comment by DonnyVerduijn** [comment]
I solved it by creating a withUseRouteHandler HOC that passes useRouteHandler as a prop to the wrapped component. useRouteHandler returns a function that takes additional props.

```js
export default function withUseRouteHandler(WrappedComponent) {
  return withRouter(({ history, location, match, ...props }) => (
    <WrappedComponent
      {...props}
      useRouteHandler={handler => {
        return handler({ match, location, history });
      }}
    />
  ));
};
```

**Comment by ruslanchek** [comment]
Did someone try this one?
https://github.com/CharlesStover/use-react-router

**Comment by PaulInglis** [comment]
> Did someone try this one?
> https://github.com/CharlesStover/use-react-router

I have.  It worked fine as long as you're not too fussy on unit testing - we struggled to get it to play ball with RTL / renderHooks

**Comment by MeiKatz** [comment]
To be honest, the router context **is** exposed, but only in a way that tells the people, "watch out, maybe we will rename it in the future, so your own code might not work in the future". You can easily import it via
```jsx
import { __RouterContext as RouterContext } from "react-router";
```

**Comment by ShamansCoding** [comment]
> I've been using this code successfully for a few months now:
> 
> ```js-jsx
> import * as React from "react";
> import { BrowserRouter, Route } from "react-router-dom";
> 
> const RouterContext = React.createContext(null);
> 
> export const HookedBrowserRouter = ({ children }) => (
>   <BrowserRouter>
>     <Route>
>       {(routeProps) => (
>         <RouterContext.Provider value={routeProps}>
>           {children}
>         </RouterContext.Provider>
>       )}
>     </Route>
>   </BrowserRouter>
> );
> 
> export function useRouter() {
>   return React.useContext(RouterContext);
> }
> ```
> 
> But either way, I don't understand why this issue is still open? Is there a complication to adding any of these variations of useRouter code?

Thanks for the solution! Easy and simple. I hope we will soon have this Hook. It'll be incredibly handy.

**Comment by ANG95** [comment]
podrias poner algo asi en tu     

> "react-router-dom": "latest",

 package.json y luego yarn install o npm i

**Comment by avindra** [comment]
Is there a discussion somewhere about the decision to expose 3 separate hooks, rather than a single `useRouter` hook?

If I'm understanding this correctly, the current hooks api will require reading the router context twice, whereas a single `useRouter` hook would've afforded us the opportunity to just consume the context in one place.

**Comment by jay-jlm** [comment]
Found this comment from @mjackson in a related PR:

> Just for posterity, we released some hooks in React Router 5.1
https://github.com/ReactTraining/react-router/releases/tag/v5.1.0
